lang/cs/Org.Apache.REEF.Bridge.Core/Grpc/DriverClientService.cs

```diff
@@ -0,0 +1,770 @@
+            DriverServiceClient driverServiceClient,
+            TcpPortProvider tcpPortProvider)
+        {
+            _injector = injector;
+            _bridgeClock = bridgeClock;
+            _driverServiceClient = driverServiceClient;
+            /*
+            bool serverStarted = false;
+            foreach (int port in tcpPortProvider)
+            {
+                try
+                {
+                    _serverPort = port;
+                    _server = new Server
+                    {
+                        Services = { DriverClient.BindService(this) },
+                        Ports = { new ServerPort("localhost", port, ServerCredentials.Insecure) }
+                    };
+                    _server.Start();
+                    serverStarted = true;
+                }
+                catch (IOException e)
+                {
+                    // keep trying
+                }
+            }
+            */
+            _server = new Server
+            {
+                Services = { DriverClient.BindService(this) },
+                Ports = { new ServerPort("localhost", 0, ServerCredentials.Insecure) }
+            };
+            _server.Start();
+            foreach (var serverPort in _server.Ports)
+            {
+                Logger.Log(Level.Info, "Server port {0}", serverPort.BoundPort);
+                _serverPort = serverPort.BoundPort;
+            }
+
+            Logger.Log(Level.Info, string.Format("Client service started on port {0}", _serverPort));
+        }
+
+        #region Driver Client Service implementation
+
+        public void Start()
+        {
+            try
+            {
+                // could throw application level exception
+                _driverBridge = _injector.GetInstance<DriverBridge>();
+                _driverServiceClient.RegisterDriverClientService("localhost", _serverPort);
+            }
+            catch (Exception ex)
+            {
+                Logger.Log(Level.Error, "Driver application creation error", ex);
+                _driverServiceClient.RegisterDriverClientService(ex);
+                _server.ShutdownAsync();
+            }
+        }
+
+        public void AwaitTermination()
+        {
+            Logger.Log(Level.Info, "Awaiting client server termination");
+            _server.ShutdownTask.Wait();
+            Logger.Log(Level.Info, "Client server terminated");
+        }
+
+        #endregion
+
+        public override async Task<Void> AlarmTrigger(AlarmTriggerInfo request, ServerCallContext context)
+        {
+            await _bridgeClock.OnNextAsync(new BridgeClock.AlarmInfo
+            {
+                AlarmId = request.AlarmId,
+                Timestamp = request.Timestamp
+            });
+            return Void;
+        }
+
+        public override Task<IdleStatus> IdlenessCheckHandler(Void request, ServerCallContext context)
+        {
+            if (IsIdle)
+            {
+                Thread.Sleep(2000);
```

Comment:
Is it worth logging here when we have hit this case? Or does this happen frequently?